Flush a fraction of the queue when we are out of space for a new kernel

jeffdaily opened the pull request:
This PR begins with the original queue flushing code, fixes bugs, and improves its performance.

Commit:
Flush a fraction of the queue when we are out of space for new kernel (#765)

When the capacity of the asyncOps vector reaches the
MAX_INFLIGHT_COMMANDS_PER_QUEUE threshold, flush only the first
fraction of the vector to overlap the cleanup process (operator
nullification and GC in removeAsyncOp()) with the ongoing GPU
computation.  The fraction size is tunable via the environment
variable HCC_QUEUE_FLUSHING_RATIO.  The default fraction size is half
of the queue.

Signed-off-by: Sarunya Pumma <sarunya.pumma@amd.com>

jeffdaily committed:
Fix queue draining code using offset. Remove GC code.

jeffdaily committed:
asyncOps compacting, better implementation

jeffdaily committed:
new env var HCC_MAX_INFLIGHT_COMMANDS_PER_QUEUE

jeffdaily committed:
drain queue starting at first found marker

jeffdaily committed:
forcefully wait if queue draining fails

scchan committed:
protect accesses to HSAQueue::asyncops with a mutex

Change-Id: I2a0cf3a97ef82479290edf833e62a55d52d68978

jeffdaily committed:
Merge branch 'clang_tot_upgrade' into clang_tot_upgrade_asyncops_drain_index

whchung commented on lib/hsa/mcwamp_hsa.cpp:
```diff
@@ -1613,30 +1624,72 @@ class HSAQueue final : public KalmarQueue
+        int first_nullptr = -1;
+        int lastWaitOp = asyncOps.size();
+        if (drainingQueue_) {
+            lastWaitOp = (lastWaitOp * QUEUE_FLUSHING_FRAC);
```
nit: maybe a static_cast here?

jeffdaily replied:
static_cast the asyncOps.size() to int?  Or static_cast `lastWaitOp*QUEUE_FLUSHING_FRAC` to int?

whchung replied:
static_cast `lastWaitOp * QUEUE_FLUSHING_FRAC`. I don't really like implicit casting for floating point numbers.

jeffdaily replied:
`QUEUE_FLUSHING_FRAC` was removed.  Logic here is updated and uses `int` calculations only.

whchung commented on lib/hsa/mcwamp_hsa.cpp:
```diff
@@ -137,6 +134,10 @@ int HCC_PROFILE_VERBOSE=0x1F;
 
+int HCC_QUEUE_FLUSHING_RATIO=50;
+
+double QUEUE_FLUSHING_FRAC;
```
nit: all calculation based on this value seems to be in `float`?

jeffdaily replied:
I believe you are suggesting to have only one variable instead of the later extra step of `QUEUE_FLUSHING_FRAC = HCC_QUEUE_FLUSHING_RATIO / 100.0`?  I could see having the user specify a value between 0 and 1, e.g., 0.5, 0.4.

Otherwise I took your comment to say use `float` instead of `double`, although I thought the default without the `f` suffix was to use `double`, so the later calculations that use `QUEUE_FLUSHING_FRAC` would be using `double` anyhow.

whchung replied:
yes, it'd be better if we could remove this global variable, or at least keep it an integer.

jeffdaily replied:
I removed `QUEUE_FLUSHING_FRAC` from my next commit.  `HCC_QUEUE_FLUSHING_RATIO` remains an `int` and is used directly when finding the index to start draining from.

whchung commented on lib/hsa/mcwamp_hsa.cpp:
```diff
@@ -1613,30 +1624,72 @@ class HSAQueue final : public KalmarQueue
+                auto asyncOp = asyncOps[i];
+                // we only drain starting at first found marker
+                if (drainingQueue_ && asyncOp->getCommandKind() != hcCommandMarker) continue;
+#if !defined(NDEBUG)
```
@jeffdaily could you clarify the rationale of this check? it seems `foundFirstValidOp` would NOT be set `true` in debug build?

whchung replied:
a deeper look suggest `foundFirstValidOp` is not really used anywhere else... should it be removed?

jeffdaily replied:
`foundFirstValidOp` and its use was present before my changes.  I was trying to preserve the original code, but I noticed that `foundFirstValidOp` was only used to trigger the code path for the `assert(sig.handle != 0)`.  Perhaps it was premature optimization, but the `assert` would be a no-op if `NDEBUG` is defined, so I tried removing the entire use of `foundFirstValidOp` if `NDEBUG` was defined.

whchung replied:
I think the logic is to ensure `asyncOp->getFuture()` won't crash later on. From `git blame` it doesn't really seem to offer any specific help. Maybe we should remove the logic here.

jeffdaily replied:
Logic removed.

whchung commented on lib/hsa/mcwamp_hsa.cpp:
```diff
@@ -1292,6 +1294,7 @@ class HSAQueue final : public KalmarQueue
     // will be waited on.
     //
     std::vector< std::shared_ptr<HSAOp> > asyncOps;
+    int asyncOps_offset;
```
comment for this important field is required.

jeffdaily replied:
Comment added in next commit.

whchung commented on lib/hsa/mcwamp_hsa.cpp:
```diff
@@ -1613,30 +1624,72 @@ class HSAQueue final : public KalmarQueue
+                    asyncOps[new_i++] = std::move(asyncOps[i]);
+                }
+                asyncOps.resize(asyncOps.size() - first_nullptr - 1);
+                asyncOps_offset += first_nullptr + 1;
```
nit: `(first_nullptr + 1)`

jeffdaily replied:
Done.

whchung commented on lib/hsa/mcwamp_hsa.cpp:
```diff
@@ -1613,30 +1624,72 @@ class HSAQueue final : public KalmarQueue
+                // We can also do better than the improved
+                // asyncOps.erase(std::remove(asyncOps.begin(), asyncOps.begin()+first_nullptr+1, nullptr), asyncOps.end());
+                // because we know that all elements from 0..first_nullptr are nullptrs,
+                // so we can avoid the redundant checks of each element against nullptr.
```
the comment seems not finished to me. could you amend the comment why erase/remove idiom not used here?

jeffdaily replied:
Amended comment.

whchung commented on lib/hsa/mcwamp_hsa.cpp:
```diff
@@ -1599,9 +1610,9 @@ class HSAQueue final : public KalmarQueue
         //
```
as the algorithm becomes much more complicated with the introduction of `asyncOps_offset` field, I think it's prudent to have a paragraph of comment to explain the algorithm better.

whchung reviewed (changes requested):
@jeffdaily overall I feel more comments are needed, plus some minor editorial changes.

whchung commented on lib/hsa/mcwamp_hsa.cpp:
```diff
@@ -77,10 +77,7 @@
-
-// threshold to clean up finished kernel in HSAQueue.asyncOps
-#define ASYNCOPS_VECTOR_GC_SIZE (2*8192)
+int HCC_MAX_INFLIGHT_COMMANDS_PER_QUEUE=(2*8192);
```
@jeffdaily I'm not very fond of this less-than-desirable coding style introduced by @bensander ... ALL-CAPITAL global variables... but we'll leave it for now

whchung commented on lib/hsa/mcwamp_hsa.cpp:
```diff
@@ -1442,19 +1445,27 @@ class HSAQueue final : public KalmarQueue
 
 
-        if (!drainingQueue_ && (asyncOps.size() >= MAX_INFLIGHT_COMMANDS_PER_QUEUE-1)) {
+        if (!drainingQueueFailed_ && !drainingQueue_ && (asyncOps.size() >= HCC_MAX_INFLIGHT_COMMANDS_PER_QUEUE-1)) {
```
a second re-read of the logic here makes it hard to distinguish which value is variable, and which is constant.

to my understanding, originally `HCC_MAX_INFLIGHT_COMMANDS_PER_QUEUE` was a constant (macro), but now it's a global variable modifiable by environment variables. it's less than desirable.

whchung replied:
.. but I think we can live with it for now as long as these values are set / accessed in a consistent fashion.

whchung commented on lib/hsa/mcwamp_hsa.cpp:
```diff
@@ -1613,30 +1624,72 @@ class HSAQueue final : public KalmarQueue
-                    if (future && future->valid()) {
-                        future->wait();
-                    }
+        for (int i = lastWaitOp-1; i >= 0;  i--) {
```
comment for this loop is necessary I think. the purpose is to find the first op which has become `nullptr`, and trigger the cleaning process in the next block?

jeffdaily replied:
Done.

whchung commented on lib/hsa/mcwamp_hsa.cpp:
```diff
@@ -1613,30 +1624,72 @@ class HSAQueue final : public KalmarQueue
+            /* we were not able to find a suitable op to wait on */
+            drainingQueueFailed_ = true;
+        }
+        else {
```
there are 3 checks in this block:
```
if (drainingQueue_ && first_nullptr >= 0) {
} else if (drainingQueue_) {
} else {
}
```

would it be better to re-organize it to:

```
if (drainingQueue_) {
  if (first_nullptr >= 0) {
  } else {
  }
} else {
}
```

so perhaps it's better comprehensible?

jeffdaily replied:
I agree, it is now more clear when reorganized as suggested.

whchung commented on lib/hsa/mcwamp_hsa.cpp:
```diff
@@ -2178,7 +2231,7 @@ class HSAQueue final : public KalmarQueue
         std::lock_guard<std::recursive_mutex> lg(qmutex);
 
-        int targetIndex = asyncOp->asyncOpsIndex();
+        int targetIndex = asyncOp->asyncOpsIndex() - asyncOps_offset;
```
any statements affected by `asyncOps_offset` deserves to have comment I think

jeffdaily replied:
Done.

whchung commented on lib/hsa/mcwamp_hsa.cpp:
```diff
@@ -2215,13 +2268,9 @@ class HSAQueue final : public KalmarQueue
-            asyncOps.erase(std::remove(asyncOps.begin(), asyncOps.end(), nullptr),
-                         asyncOps.end());
-        }
+        /* We used to garbage collect nullptr entries from asyncOps here.
```
nit: use `//` in all comments, both single-line and multi-line.

jeffdaily replied:
Done.

whchung commented on lib/hsa/mcwamp_hsa.cpp:
```diff
@@ -1442,19 +1445,27 @@ class HSAQueue final : public KalmarQueue
+            }
         }
-        op->asyncOpsIndex(asyncOps.size());
+        op->asyncOpsIndex(asyncOps.size()+asyncOps_offset);
```
any statement be associated with `asyncOps_offset` deserves to have comment on the effect brought by this important field.

jeffdaily replied:
Done.

jeffdaily committed:
Address review comments.

- Additional code comments.
- Clearer if/else logic.
- Removed debug code and variable.

jeffdaily reviewed (commented):
New commit addressing comments has been pushed.

whchung reviewed (approved):
Thanks for the revision! LGTM

scchan commented on lib/hsa/mcwamp_hsa.cpp:
```diff
@@ -1613,30 +1631,77 @@ class HSAQueue final : public KalmarQueue
+                // wait on valid futures only
+                std::shared_future<void>* future = asyncOp->getFuture();
+                if (future && future->valid()) {
+                    future->wait();
```
We are holding the lock to this asyncOps and wait so this may cause a potential deadlock here, need to take a closer look

jeffdaily replied:
`future->wait()` happens in the original code for `HSAQueue::wait()`, so this deadlock might not be specific to my PR.   I'm looking at this code closely now, and indeed the thread holding the lock for `asyncOps` will trigger `hsa_signal_wait_scacquire` because of the `future->wait()`.  If queue stealing occurs by another thread while the signal wait is blocking, could this cause the deadlock?

scchan commented:
superseded by PR1046
